From de5f44d5d82f99c44f65cbcac297bbfe8149ce8a Mon Sep 17 00:00:00 2001
From: Nai Xia <nai.xia@gmail.com>
Date: Thu, 18 Oct 2012 20:20:42 +0800
Subject: [PATCH] Mute page allocation failure warning message

---
 mm/uksm.c |    2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/mm/uksm.c b/mm/uksm.c
index 967c755..81400af 100644
--- a/mm/uksm.c
+++ b/mm/uksm.c
@@ -2901,7 +2901,7 @@ struct rmap_list_entry *get_rmap_list_entry(struct vma_slot *slot,
 		if (!need_alloc)
 			return NULL;
 
-		page = alloc_page(GFP_KERNEL | __GFP_ZERO);
+		page = alloc_page(GFP_KERNEL | __GFP_ZERO | __GFP_NOWARN);
 		if (!page)
 			return NULL;
 
-- 
1.7.9.5

