diff -ru boot/initrd/fstab.cardhu b/boot/initrd/fstab.cardhu
--- a/boot/initrd/fstab.cardhu	2012-12-17 17:32:45.000000000 +0400
+++ b/boot/initrd/fstab.cardhu	2013-02-10 03:17:40.704776814 +0400
@@ -3,6 +3,7 @@
 # The filesystem that contains the filesystem checker binary (typically /system) cannot
 # specify MF_CHECK, and must come before any filesystems that do specify MF_CHECK
 
-/dev/block/mmcblk0p1	/system	ext4	noatime,noauto_da_alloc,nodiratime,nobh,ro	wait
-/dev/block/mmcblk0p2	/cache	ext4	noatime,noauto_da_alloc,nosuid,nodev,nodiratime,nobh,errors=panic	wait
-/dev/block/mmcblk0p8	/data	ext4	noatime,noauto_da_alloc,nosuid,nodev,nodiratime,nobh,errors=panic	wait,encryptable=footer
+/dev/block/mmcblk0p1	/system	ext4	noatime,noauto_da_alloc,nodiratime,barrier=0,nobh,ro	wait
+/dev/block/mmcblk0p2	/cache	ext4	noatime,noauto_da_alloc,nosuid,nodev,nodiratime,barrier=0,nobh,errors=panic	wait
+/dev/block/mmcblk1p2	/data	ext4	noatime,noauto_da_alloc,nosuid,nodev,nodiratime,barrier=0,nobh,errors=panic	wait,encryptable=footer
+/dev/block/mmcblk0p8	/data2	ext4	noatime,noauto_da_alloc,nosuid,nodev,nodiratime,barrier=0,nobh,errors=panic	wait,encryptable=footer
diff -ru boot/initrd/init.cardhu.rc b/boot/initrd/init.cardhu.rc
--- a/boot/initrd/init.cardhu.rc	2012-12-17 17:32:45.000000000 +0400
+++ b/boot/initrd/init.cardhu.rc	2013-02-10 03:17:40.708776814 +0400
@@ -13,8 +13,9 @@
 
     # for backwards compatibility
     export EXTERNAL_STORAGE /storage/sdcard0
-    export SECONDARY_STORAGE /storage/sdcard1
+    export SECONDARY_STORAGE /storage/sdcardi
     mkdir /storage 0550 system sdcard_r
+    mkdir /storage/sdcardi 0000 system system
     mkdir /storage/sdcard0 0000 system system
     mkdir /storage/sdcard1 0000 system system
     symlink /storage/sdcard0 /sdcard
@@ -32,6 +33,9 @@
 on fs
     setprop ro.crypto.tmpfs_options size=128m,mode=0771,uid=1000,gid=1000
     setprop ro.crypto.umount_sd false
+    mkdir /data2 0000 system system
+    symlink /storage/sdcardi /mnt/sdcard_internal
+    symlink /storage/sdcardi /sdcardi
     mount_all /fstab.cardhu
     mount tmpfs tmpfs /Removable mode=0755,gid=1000
 
@@ -328,6 +332,8 @@
 service sdcard /system/bin/sdcard /data/media 1023 1023
     class late_start
 
+service sdcardi /system/bin/sdcardi /data2/media 1023 1023
+    class late_start
 # bugreport is triggered by the VOLUME-DOWN and VOLUME-UP keys
 service bugreport /system/bin/bugmailer.sh -b /system/media/audio/ui/Dock.ogg -e /system/media/audio/ui/Undock.ogg
     class main
