diff -urN a/drivers/ril/ril.c b/drivers/ril/ril.c
--- a/drivers/ril/ril.c	2012-09-19 12:45:36.000000000 +0200
+++ b/drivers/ril/ril.c	2012-10-23 14:03:58.000000000 +0200
@@ -9,7 +9,7 @@
 #include <linux/uaccess.h>
 #include <linux/workqueue.h>
 #include <linux/wakelock.h>
-#include <../../arch/arm/mach-tegra/include/mach/board-cardhu-misc.h>
+#include "../../arch/arm/mach-tegra/include/mach/board-cardhu-misc.h"
 
 #include "pm-irq.h"
 #include "ril.h"
